lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs

```diff
@@ -254,26 +272,143 @@ instance ((PersistPublicKey (key 'AccountK)), Typeable chain)
+    let updateKey (ix, _) = (ix, Used)
+    in case Map.lookup keyHash indexedHashKeys of
+        Nothing -> (Nothing, pool)
+        Just (ix, Unused) ->
+            ( Just ix
+            , unsafeVerificationKeyPool accXPub g
+                (Map.adjust updateKey keyHash indexedHashKeys) scripts)
+        Just (ix, Used) ->
+            ( Just ix
+            , unsafeVerificationKeyPool accXPub g indexedHashKeys scripts)
+
+updateKnownScripts
+    :: (Map ScriptHash [Index 'Soft 'ScriptK] -> Map ScriptHash [Index 'Soft 'ScriptK])
+    -> VerificationKeyPool k
+    -> VerificationKeyPool k
+updateKnownScripts trKnownScripts (VerificationKeyPool accXPub g indexedHashKeys knownScripts) =
+    let knownScripts' = trKnownScripts knownScripts
+    in unsafeVerificationKeyPool accXPub g indexedHashKeys knownScripts'
+
+-- | Create a new VerificationKey pool.
+newVerificationKeyPool
+    :: (SoftDerivation k, WalletKey k)
+    => k 'AccountK XPub
+    -> AddressPoolGap
+    -> VerificationKeyPool k
+newVerificationKeyPool accXPub addrPoolGap =
+    let startVerKeyMap =
+            Map.fromList $ map (createPristineMapEntry accXPub) $
+            L.take (fromEnum addrPoolGap) $ L.iterate succ minBound
+    in unsafeVerificationKeyPool accXPub addrPoolGap startVerKeyMap Map.empty
+
+-- | Create a VerificationKey pool from the all ingredients.
+-- There is no validation and it is unsafe way.
+unsafeVerificationKeyPool
+    :: k 'AccountK XPub
+    -> AddressPoolGap
+    -> Map KeyHash (Index 'Soft 'ScriptK, AddressState)
+    -> Map ScriptHash [Index 'Soft 'ScriptK]
+    -> VerificationKeyPool k
+unsafeVerificationKeyPool accXPub poolGap vkPoolMap knownScripts = VerificationKeyPool
+    { verPoolAccountPubKey = accXPub
+    , verPoolGap = poolGap
+    , verPoolIndexedKeys = vkPoolMap
+    , verPoolKnownScripts = knownScripts
+    }
+
+createPristineMapEntry
+    :: (SoftDerivation k, WalletKey k)
+    => k 'AccountK XPub
+    -> Index 'Soft 'ScriptK
+    -> (KeyHash, (Index 'Soft 'ScriptK, AddressState))
+createPristineMapEntry accXPub ix =
+    ( hashVerificationKey $ deriveVerificationKey accXPub ix
+    , (ix, Unused) )
+
+-- | The extension to the VerificationKey pool is done by adding next adjacent
+-- indices and their corresponding public keys, marking them as Unused.
+-- The number of added entries is determined by pool gap, and the start index is
+-- the next to the the specified index.
+extendVerificationKeyPool
+    :: (SoftDerivation k, WalletKey k)
+    => Index 'Soft 'ScriptK
+    -> VerificationKeyPool k
+    -> VerificationKeyPool k
+extendVerificationKeyPool ix pool
+    | isOnEdge  = pool { verPoolIndexedKeys = Map.union (verPoolIndexedKeys pool) next }
+    | otherwise = pool
+  where
+    edge = Map.size (verPoolIndexedKeys pool)
+    isOnEdge = edge - fromEnum ix <= fromEnum (verPoolGap pool)
+    next =
+        if ix == maxBound then mempty
+        else
```

Comment:
small stylistic suggestion: guards are also nice for such cases:

```hs
next
    | ix == maxBound = mempty
    | otherwise      = ...
```

Reply:
done in https://github.com/input-output-hk/cardano-wallet/pull/2308/commits/ab8567640737c089cb9f15c4be3cfce6dbcadb1a